server/server.go

```diff
@@ -177,27 +187,31 @@ func (h *handler) handle(pkt *protocol.Packet) {
 	}
 	h.tokens <- struct{}{}
 	h.mtx.Lock()
-	_, err := respPkt.WriteTo(h.conn)
-	h.mtx.Unlock()
+	defer h.mtx.Unlock()
+	err := h.conn.SetWriteDeadline(time.Now().Add(h.timeout))
 	if err != nil {
-		log.Errorf("connection %v: error in writing response %v", h.name, err)
-		h.conn.Close() // ignoring error: what can we do?
-		h.s.mtx.Lock()
-		delete(h.s.listeners[h.listener], h.conn)
-		h.s.mtx.Unlock()
-		logConnFailure()
+		h.closeWithWritingErr(err)
+		return // we lose the metrics, but this is an exceptional case.
```

Comment:
Definitely an exceptional case, but you could handle it for all cases by moving the metrics logging to a defer:
```go
h.mtx.Lock()
defer h.mtx.Unlock()

defer func() {
	logRequestExecDuration(pkt.Operation.Opcode, start, resp.op.ErrorVal())
	logRequestTotalDuration(pkt.Operation.Opcode, start, resp.op.ErrorVal())
}()

err := h.conn.SetWriteDeadline(time.Now().Add(h.timeout))
if err != nil {
	h.closeWithWritingErr(err)
	return
}

_, err = respPkt.WriteTo(h.conn)
if err != nil {
	h.closeWithWritingErr(err)
	return
}
```

Reply:
+1 It also seemed to me that having metrics when an exceptional case is encountered would be even more necessary